### PR TITLE
fix(calendar): no setting of date if same date is already set

### LINKED ISSUE #13850: Binding 2 calendars to the same value makes the app freeze
## Description  
Describe the issue.

 * igniteui-angular version: ^17.0.0
 * browser: Chrome/FF

## Steps to reproduce  

1.
```
@Component({
  selector: 'card-sample',
  templateUrl: './card-sample.component.html'
})
export class CardSampleComponent {
  public date = new Date('2024-01-30T00:00');
}
```
```
  <igx-calendar [(ngModel)]="date"></igx-calendar>
  <igx-calendar [(ngModel)]="date"></igx-calendar>
```

## Result  
Everything freezes, high CPU usage

## Expected result  
Display both calendars bound to the same value

## Attachments  
Stackblitz: https://stackblitz.com/edit/s81ez8?file=src%2Fapp%2Fcalendar-sample-5%2Fcalendar-sample-5.component.html)

### PR DESCRIPTION
**Pull request by deyvidnenchev**
Closes #13850

### Additional information (check all that apply):
 - [x] Bug fix
 - [ ] New functionality
 - [ ] Documentation
 - [ ] Demos
 - [ ] CI/CD

### Checklist:
 - [x] All relevant tags have been applied to this PR
 - [ ] This PR includes unit tests covering all the new code ([test guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Test-implementation-guidelines-for-Ignite-UI-for-Angular))
 - [ ] This PR includes API docs for newly added methods/properties ([api docs guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Documentation-Guidelines))
 - [ ] This PR includes `feature/README.MD` updates for the feature docs
 - [ ] This PR includes general feature table updates in the root `README.MD`
 - [ ] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [ ] This PR contains breaking changes
 - [ ] This PR includes `ng update` migrations for the breaking changes ([migrations guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Update-Migrations))
 - [ ] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

**Review by GrblackStar — APPROVED** [review]

File: projects/igniteui-angular/src/lib/calendar/calendar-base.ts [hunk 1]
```diff
@@ -649,8 +649,11 @@ export class IgxCalendarBaseDirective implements ControlValueAccessor {
      * @hidden
      */
     private selectSingle(value: Date) {
-        this.selectedDates = this.getDateOnly(value);
-        this._onChangeCallback(this.selectedDates);
+        const selectedDate = this.selectedDates as Date;
+        if(!selectedDate || selectedDate.getTime() !== value.getTime()){
```
Comment (kdinev) [inline projects/igniteui-angular/src/lib/calendar/calendar-base.ts:653]:
```suggestion
        if (!selectedDate || selectedDate.getTime() !== value.getTime()) {
```

**Review by kdinev — CHANGES_REQUESTED** [review]

**Review by kdinev — DISMISSED** [review]

File: projects/igniteui-angular/src/lib/calendar/calendar-base.ts [hunk 2]
```diff
@@ -649,8 +649,11 @@ export class IgxCalendarBaseDirective implements ControlValueAccessor {
      * @hidden
      */
     private selectSingle(value: Date) {
-        this.selectedDates = this.getDateOnly(value);
-        this._onChangeCallback(this.selectedDates);
+        const selectedDate = this.selectedDates as Date;
+        if (!selectedDate || selectedDate.getTime() !== value.getTime()) {
```
Comment (pmoleri) [inline projects/igniteui-angular/src/lib/calendar/calendar-base.ts:653]: This cast is not correct. The fact that the next line checks for `null/undefined` means that it should have been at least `Date | null | undefined`, but what worries me more is if it's not a Date object the `.getTime()` in the next line will crash.

I think adding a compare method will be safer:
```ts
private selectionEquals(a: unknown, b: unknown): boolean {
   if (a instanceof Date && b instanceof Date) {
       return a.getTime() === b.getTime();
    } else if (Array.isArray(a) && Array.isArray(b)) {
        return a.length === b.length && a.every((v, i) => this.selectionEquals(v, b[i]));
    } else {
        return a === b;
    }
}
```

Also, by looking at the code, it's very probable that the same issue can be reproduced with arrays in multiple selection mode

**Review by GrblackStar — APPROVED** [review]

**Review by GrblackStar — APPROVED** [review]

**Comment by kdinev** [comment]
@pmoleri Do you reject the fix and should I wait for more changes before merging?

**Comment by hanastasov** [comment]
Actually what @pmoleri highlighted is logged as a separate issue. This PR also waits for the same fix for range selection calendar. I will set the proper status. still in development.

**Pull request merged**